pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartitionUpsertOffHeapMetadataManager.java

```diff
@@ -0,0 +1,253 @@
+              new RecordLocationRef(segmentId, recordInfo.getDocId(), recordInfo.getComparisonValue());
+          updateRecordInfo(primaryKeyId, newRecordLocationRef);
+        } else {
+          //updateRecordInfo(primaryKeyId, currentRecordLocation);
+        }
+      } else {
+        // New primary key
+        validDocIds.add(recordInfo.getDocId());
+        RecordLocationRef newRecordLocationRef =
+            new RecordLocationRef(segmentId, recordInfo.getDocId(), recordInfo.getComparisonValue());
+        updateRecordInfo(primaryKeyId, newRecordLocationRef);
+      }
+    }
+  }
+
+  private int getPrimaryKeyId(RecordInfo recordInfo) {
+    byte[] pk = recordInfo.getPrimaryKey().asBytes();
+    int primaryKeyId = _bytesOffHeapMutableDictionary.index(pk);
+    return primaryKeyId;
+  }
+
+  /**
+   * Updates the upsert metadata for a new consumed record in the given consuming segment.
+   */
+  public void addRecord(IndexSegment segment, RecordInfo recordInfo) {
+    ThreadSafeMutableRoaringBitmap validDocIds = Objects.requireNonNull(segment.getValidDocIds());
+    int segmentId = _segmentToSegmentIdMap.computeIfAbsent(segment, (segmentObj) -> {
+      Integer newSegmentId = _segmentId.incrementAndGet();
+      _segmentIdToSegmentMap.put(newSegmentId, segment);
+      return newSegmentId;
+    });
+
+    int primaryKeyId = getPrimaryKeyId(recordInfo);
+
+    RecordLocationRef value = getRecordInfo(primaryKeyId);
+
+    if (value != null) {
+      RecordLocationRef currentRecordLocation = value;
+      if (recordInfo.getComparisonValue().compareTo(currentRecordLocation.getComparisonValue()) >= 0) {
+        IndexSegment currentSegment = (IndexSegment) _segmentIdToSegmentMap.get(currentRecordLocation.getSegmentRef());
+        int currentDocId = currentRecordLocation.getDocId();
+        if (segment == currentSegment) {
+          validDocIds.replace(currentDocId, recordInfo.getDocId());
+        } else {
+          Objects.requireNonNull(currentSegment.getValidDocIds()).remove(currentDocId);
+          validDocIds.add(recordInfo.getDocId());
+        }
+        RecordLocationRef newRecordLocationRef =
+            new RecordLocationRef(segmentId, recordInfo.getDocId(), recordInfo.getComparisonValue());
+        updateRecordInfo(primaryKeyId, newRecordLocationRef);
+      } else {
+        updateRecordInfo(primaryKeyId, currentRecordLocation);
+      }
+    } else {
+      validDocIds.add(recordInfo.getDocId());
+      RecordLocationRef recordLocationRef =
+          new RecordLocationRef(segmentId, recordInfo.getDocId(), recordInfo.getComparisonValue());
+      updateRecordInfo(primaryKeyId, recordLocationRef);
+    }
+  }
+
+  /**
+   * Removes the upsert metadata for the given immutable segment. No need to remove the upsert metadata for the
+   * consuming segment because it should be replaced by the committed segment.
+   */
+  public void removeSegment(IndexSegment segment) {
+    throw new NotImplementedException();
```

Comment:
The 2 data structures we use here cannot support remove operation

Reply:
I was planning on putting a dummy value for the key whever it is deleted. Other than that yeah there's no way.